File: 279. Perfect Squares/279.go
package leetcode

import(
	"math"
)

func numSquares(n int) int {
	mylist := []int{n}
	step := 0

	for{
		step ++
		tmpDict := make(map[int]int, 0)
		for len(mylist)>0{
			n := mylist[0]
			mylist = mylist[1:]
			n1 := int(math.Sqrt(float64(n)))+1
			for i:=1; i<n1; i++{
				cha := n - i*i
				if cha == 0{
					return step
				}
				tmpDict[cha] = 1
			}
		}
		for i := range tmpDict{
			mylist = append(mylist, i)
		}
	}
}

File: 279. Perfect Squares/_279.py
def solution(n: int) -> int:
        mylist = [n]
        step = 0

        while True:
            step += 1
            tmp = set()
            while mylist:
                n = mylist.pop()
                for i in range(1, int(n**0.5)+1):
                    cha = n-i**2
                    if cha == 0:
                        return step
                    else:
                        tmp.add(cha)
            mylist = list(tmp)

        return step